modules/yieldlabBidAdapter.js

```diff
@@ -221,6 +230,35 @@ function createTargetingString (obj) {
   return str.join('&')
 }
 
+/**
+ * Creates a string out of a schain object
+ * @param {Object} schain
+ * @returns {String}
+ */
+function createSchainString (schain) {
+  let ver = schain.ver ? schain.ver : ''
+  let complete = schain.complete ? schain.complete : ''
+  let nodes = []
+  for (let i = 0; i < schain.nodes.length; i++) {
+    let node = []
+    node.push(schain.nodes[i].asi, schain.nodes[i].sid, schain.nodes[i].hp, schain.nodes[i].rid, schain.nodes[i].name, schain.nodes[i].domain, schain.nodes[i].ext)
+    let filteredAndEncodedNode = node.map(x => x === undefined ? '' : encodeURIComponentWithBangIncluded(x))
```

Comment:
Run this function on each item when it is pushed into the array. No reason to push the items and then immediately loop over them.

Reply:
I changed this altogether to not use arrays anymore.

Reply:
Using arrays is fine. I only meant that it could be simplified. What you switched to works, but since the code now repeats the same line for each key, you could define an array of the key strings and map over that for each schain node.

```
const keys = ['asi', 'sid', 'hp', 'rid', 'name', 'domain', 'ext'];
const nodesString = schain.nodes.reduce((acc, node) => {
    return acc += `!${keys.map(key => node[key] ? encodeURIComponentWithBangIncluded(node[key]) : '').join()}`
}, '');
```

Reply:
Thanks for the input and help here @msm0504!  I did not know about the reduce method till now. I just used your code (only had to include the join separator `,`).